base/mapiterator.jl

```diff
@@ -0,0 +1,175 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+
+"""
+    MapIterator(f, c...) -> iterator
+
+`collect(MapIterator(f, c...))` is equivalent to `map(f, c...)`
+"""
+
+type MapIterator
+    f::Union{Function,Type}
+    arg_itr
+end
+
+MapIterator(f, c...) = MapIterator(f, zip(c...))
+
+
+start(itr::MapIterator) = start(itr.arg_itr)
+
+done(itr::MapIterator, state) = done(itr.arg_itr, state)
+
+function next(itr::MapIterator, state)
+    args, state = next(itr.arg_itr, state)
+    itr.f(args...), state
+end
+
+
+
+"""
+    AsyncMapIterator(f, results, c...; ntasks=100) -> iterator
+
+Apply f to each element of c using at most 100 asynchronous tasks.
+For multiple collection arguments, apply f elementwise.
+Results are asynchronously stored in "results" collection
+(the iterator returns nothing).
+
+Note: `for task in AsyncMapIterator(f, results, c...)` end is equivalent to
+`map!(f, results, c...)`
+"""
+
+type AsyncMapIterator
+    f
+    results
+    arg_enum::Enumerate
+    ntasks::Int
+end
+
+function AsyncMapIterator(f, results, c...; ntasks=0)
+    if ntasks == 0
+        ntasks = 100
+    end
+    AsyncMapIterator(f, results, enumerate(zip(c...)), ntasks)
+end
+
+
+type AsyncMapState
+    enum_state
+    active_count::Int
+    task_done::Condition
+    done::Bool
+end
+
+
+# Busy if the maximum number of concurrent tasks is running.
+function isbusy(itr::AsyncMapIterator, state::AsyncMapState)
+    state.active_count == itr.ntasks
+end
+
+
+# Wait for @async task to end.
+wait(state::AsyncMapState) = wait(state.task_done)
+
+
+# Open a @sync block and initialise iterator state.
+function start(itr::AsyncMapIterator)
+    Base.sync_begin()
+    AsyncMapState(start(itr.arg_enum),  0, Condition(), false)
+end
+
+# Close @sync block when iterator is done.
+function done(itr::AsyncMapIterator, state::AsyncMapState)
+    if !state.done && done(itr.arg_enum, state.enum_state)
+        state.done = true
+        Base.sync_end()
+    end
+    return state.done
+end
+
+function next(itr::AsyncMapIterator, state::AsyncMapState)
+
```

Comment:
IMO this is less readable, not serving any purpose and the majority of the rest of the code in base doesn't do this. That sentence is mostly referring to spaces after commas.